### PR TITLE
Added gradient() evaluation to VectorFunctionFromTensorFunction

### PR DESCRIPTION
**Pull request by AbbBallout**
The gradient() withing the VectorFunctionFromTensorFunction class was not implemented. 
I added it. The functions gradient(), gradient_value list() .... were implemented by copying code from Function class. 
I tested the code by returning the H1 seminorm of a linear elasticity example and I got the right convergence rates.

### CODE REVIEW

File: include/deal.II/base/function.h [hunk 1]
```diff
@@ -1148,6 +1148,64 @@ class VectorFunctionFromTensorFunction : public Function<dim, RangeNumberType>
...
+   *
+   * The outer loop over <tt>gradients</tt> is over the points in the list,
+   * the inner loop over the different components of the function.
+   */
+  virtual void
+  vector_gradient_list(const std::vector<Point<dim>> &points,
+                       std::vector<std::vector<Tensor<1, dim, RangeNumberType>>>
+                         &gradients) const override;
+
+  /**
```
Comment (tjhei) [inline include/deal.II/base/function.h:1205]: I think you don't want to include this comment here.

**Review by tjhei — APPROVED** [review]
Thank you for your contribution!

This looks very good already. Ideally, we would have a small test for the new functionality. Have you looked at the Testsuite before? One of us is happy to help.

File: include/deal.II/base/function.h (hunk 1 above)
Comment (AbbBallout) [reply include/deal.II/base/function.h:1205]: I found a test at base/functions_08.cc. 
I believe I should figure how that works and create something similar. Right? 
I'll take the help of course. How can I build and run one single test?

Comment (tjhei) [reply include/deal.II/base/function.h:1205]: Yes, make a copy of an existing test. You can follow the guide at https://www.dealii.org/developer/developers/testsuite.html#layoutaddtests
This page has a lot of other information to get you started.

Comment (AbbBallout) [reply include/deal.II/base/function.h]: I made a test called functions_08_gradient based on functions_08. 

0) These test now look a lot like one another. Is it a good/bad idea to make them into one? 
1) Is this good enough or should I make it better? 
2) Does the cli get triggered everytime I push a commit here? Should I be careful about that?  
3) Once this gets merged with master deal, do all the commits that I did show up in the master branch or will my changes show as a one big commit?
4) Some of the pull-requests are getting tags (infrastructure, build system, bug). Is it me who should set that?

**Review by tjhei — APPROVED** [review]
This is your first pull request, so we need to manually approve the tests to run. You should now see the status of the various testers and they hopefully will turn green. 😉

Copying a test like this is totally fine.

**Comment by tjhei** [comment]
All your individual commits will become part of the deal.II history when merged. This is why we typically try to have pull requests with a small number of clean commits. We typically achieve this by letting the person who opened the PR squash the commits. Feel free to do so, but 3 commits is also fine with me.

**Comment by AbbBallout** [comment]
I'll leave the commits that way.

**Review by masterleinad — COMMENTED** [review]
Would you mind adding a changelog entry (see the ones in `doc/news/changes/minor/`)?

**Comment by AbbBallout** [comment]
@masterleinad of course. I have tagged this as an improvement.

**Review by masterleinad — APPROVED** [review]

File: doc/news/changes/minor/20230730AbbasBallout [hunk 2]
```diff
@@ -0,0 +1,4 @@
+Improvement: Added gradient() implimentation to VectorFunctionFromTensorFunction in 
+function.h file. tests/base/function/functions_08_gradient.cc was also added.
```
Comment (masterleinad) [inline doc/news/changes/minor/20230730AbbasBallout:2]:
```suggestion
Improvement: Added gradient() implementation to VectorFunctionFromTensorFunction in 
function.h file.```

File: doc/news/changes/minor/20230730AbbasBallout [hunk 3]
```diff
@@ -0,0 +1,4 @@
+Improvement: Added gradient() implementation to VectorFunctionFromTensorFunction in 
+function.h file.```
```
Comment (masterleinad) [inline doc/news/changes/minor/20230730AbbasBallout:2]: Almost. 🙂 
```suggestion
Improvement: Added gradient() implementation to VectorFunctionFromTensorFunction in 
function.h file.
```

**Review by masterleinad — APPROVED** [review]

**Comment by AbbBallout** [comment]
The indent job failed because I didn't have my full name set here. I fixed it.

**Comment by masterleinad** [comment]
> The indent job failed because I didn't have my full name set here. I fixed it.

You need to fix it for all commits.

**Comment by masterleinad** [comment]
> Can we delete this pull request and start a new one if this doesn't work?

Just do whatever you would do for a new pull request and then force-push to this branch.

**Comment by AbbBallout** [comment]
My name is correctly set now. Do we keep this at 12 commits? Is it okay? Can I do something about it?

**Comment by masterleinad** [comment]
The indentation check is still failing. You really have to update your name for every single commit. Ultimately, you could just squash all commits and make sure this last commit is conforming.

**Comment by masterleinad** [comment]
> I thought it worked when I first checked.
> I need some help with this.

Do something like
```
git rebase -i upstream/master
```
and replace all `pick` by `squash` (expect the first commit), also see https://www.git-tower.com/learn/git/faq/git-squash.
and then force-push to this branch via
```
git push -f origin VectorFunctionFromTensorFunction_gradient
```
(assuming that your fork corresponds to the `origin` remote and that deal.II repository is your `upstream` remote).

**Comment by AbbBallout** [comment]
Thank you. 
This should work now.

**Comment by AbbBallout** [comment]
Is this failed check normal?

File: doc/news/changes/minor/20230730AbbasBallout [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+=======
+Improvement: Added gradient() implementation to VectorFunctionFromTensorFunction in 
+<<<<<<< HEAD
+function.h file.```
+>>>>>>> 6c21023869 (Update doc/news/changes/minor/20230730AbbasBallout)
+=======
+function.h file.
+>>>>>>> f108d559f8 (Update doc/news/changes/minor/20230730AbbasBallout)
+<br>
+(Abbas Ballout, 2023/07/30)
```
Comment (masterleinad) [inline doc/news/changes/minor/20230730AbbasBallout:18]: What happened here?

Comment (AbbBallout) [reply doc/news/changes/minor/20230730AbbasBallout]: I pushed before checking the files. 
How many more chances do I have?

Comment (masterleinad) [reply doc/news/changes/minor/20230730AbbasBallout]: As many as you need. 🙂

**Comment by AbbBallout** [comment]
Thank you @tjhei and @masterleinad.
It was a nice learning experience for me as well.
Do I close the pull request?

**Comment by masterleinad** [comment]
> Do I close the pull request?

No, I merged it. 🙂

**Comment by tjhei** [comment]
Thanks again, @AbbBallout . Hopefully not your last contribution. :-)

**Pull request merged**